### PR TITLE
Basic trainers fixes

### LINKED ISSUE #64: Investigate the content of statistics, rethink the desired behavior of online/offline trainers
When running grid trainer with "weird" terminal conditions:
        # Set the same terminal conditions.
        terminal_conditions:
            loss_stop: 1.0e-5
            episode_limit: 100
            epochs_limit: 10 

I received even weirder results, i.e.:
- content of training_statistics.csv:

loss | episode | epoch | acc | batch_size
-- | -- | -- | -- | --
2.3023982 | 0 | 0 | 0.0625 | 64
2.42056847 | 1 | 0 | 0.09375 | 64
2.31480336 | 2 | 0 | 0.015625 | 64
2.30580044 | 3 | 0 | 0.125 | 64
2.28870821 | 4 | 0 | 0.109375 | 64
2.29206491 | 5 | 0 | 0.140625 | 64
2.29954505 | 6 | 0 | 0.109375 | 64
2.31241441 | 7 | 0 | 0.0625 | 64
2.2548635 | 8 | 0 | 0.1875 | 64
2.27464652 | 9 | 0 | 0.078125 | 64
2.24816489 | 10 | 0 | 0.171875 | 64
2.22160149 | 11 | 0 | 0.140625 | 64
2.14928031 | 12 | 0 | 0.171875 | 64
2.01351118 | 13 | 0 | 0.125 | 64
1.9892391 | 14 | 0 | 0.203125 | 64
1.78259158 | 15 | 0 | 0.375 | 64
1.76220071 | 16 | 0 | 0.40625 | 64
1.59635079 | 17 | 0 | 0.5625 | 64
1.24760115 | 18 | 0 | 0.59375 | 64
1.52464759 | 19 | 0 | 0.578125 | 64
1.02153337 | 20 | 0 | 0.6875 | 64
1.13793755 | 21 | 0 | 0.65625 | 64
1.04477477 | 22 | 0 | 0.703125 | 64
0.9281581 | 23 | 0 | 0.65625 | 64
0.8872776 | 24 | 0 | 0.75 | 64
0.96978921 | 25 | 0 | 0.71875 | 64
0.54203308 | 26 | 0 | 0.84375 | 64
0.71089166 | 27 | 0 | 0.828125 | 64
0.87194252 | 28 | 0 | 0.78125 | 64
0.74255377 | 29 | 0 | 0.796875 | 64
0.97673011 | 30 | 0 | 0.71875 | 64
0.89663988 | 31 | 0 | 0.671875 | 64
0.52516961 | 32 | 0 | 0.859375 | 64
0.67374974 | 33 | 0 | 0.6875 | 64
1.07340407 | 34 | 0 | 0.640625 | 64
0.65307564 | 35 | 0 | 0.71875 | 64
0.62181377 | 36 | 0 | 0.796875 | 64
0.53182065 | 37 | 0 | 0.84375 | 64
0.71876705 | 38 | 0 | 0.8125 | 64
0.69108903 | 39 | 0 | 0.8125 | 64
0.64931148 | 40 | 0 | 0.84375 | 64
0.64401007 | 41 | 0 | 0.765625 | 64
0.53108335 | 42 | 0 | 0.78125 | 64
0.47211882 | 43 | 0 | 0.8125 | 64
0.42510599 | 44 | 0 | 0.859375 | 64
0.53872186 | 45 | 0 | 0.78125 | 64
0.47975114 | 46 | 0 | 0.875 | 64
0.42682296 | 47 | 0 | 0.84375 | 64
0.68501329 | 48 | 0 | 0.84375 | 64
0.51802105 | 49 | 0 | 0.84375 | 64
0.42391279 | 50 | 0 | 0.890625 | 64
0.54398292 | 51 | 0 | 0.8125 | 64
0.43966454 | 52 | 0 | 0.828125 | 64
0.41222006 | 53 | 0 | 0.859375 | 64
0.30380589 | 54 | 0 | 0.890625 | 64
0.28526509 | 55 | 0 | 0.890625 | 64
0.3890301 | 56 | 0 | 0.890625 | 64
0.3726145 | 57 | 0 | 0.859375 | 64
0.37899232 | 58 | 0 | 0.84375 | 64
0.31090876 | 59 | 0 | 0.90625 | 64
0.29964575 | 60 | 0 | 0.890625 | 64
0.29754484 | 61 | 0 | 0.875 | 64
0.30940181 | 62 | 0 | 0.90625 | 64
0.28904665 | 63 | 0 | 0.90625 | 64
0.28892154 | 64 | 0 | 0.9375 | 64
0.28293437 | 65 | 0 | 0.890625 | 64
0.28884795 | 66 | 0 | 0.9375 | 64
0.27016857 | 67 | 0 | 0.890625 | 64
0.38757008 | 68 | 0 | 0.921875 | 64
0.24764507 | 69 | 0 | 0.921875 | 64
0.25606325 | 70 | 0 | 0.890625 | 64
0.48922622 | 71 | 0 | 0.84375 | 64
0.2977196 | 72 | 0 | 0.890625 | 64
0.3917419 | 73 | 0 | 0.921875 | 64
0.19252293 | 74 | 0 | 0.9375 | 64
0.39461273 | 75 | 0 | 0.875 | 64
0.28725618 | 76 | 0 | 0.859375 | 64
0.24857962 | 77 | 0 | 0.921875 | 64
0.22327447 | 78 | 0 | 0.9375 | 64
0.41391894 | 79 | 0 | 0.859375 | 64
0.19850856 | 80 | 0 | 0.921875 | 64
0.30375871 | 81 | 0 | 0.890625 | 64
0.38144702 | 82 | 0 | 0.890625 | 64
0.29862314 | 83 | 0 | 0.921875 | 64
0.16170724 | 84 | 0 | 0.953125 | 64
0.25888351 | 85 | 0 | 0.953125 | 64
0.17384183 | 86 | 0 | 0.953125 | 64
0.24882084 | 87 | 0 | 0.953125 | 64
0.20304871 | 88 | 0 | 0.921875 | 64
0.354817 | 89 | 0 | 0.9375 | 64
0.12355755 | 90 | 0 | 0.96875 | 64
0.20728019 | 91 | 0 | 0.921875 | 64
0.17258625 | 92 | 0 | 0.921875 | 64
0.16974132 | 93 | 0 | 0.953125 | 64
0.37275705 | 94 | 0 | 0.90625 | 64
0.09402215 | 95 | 0 | 0.96875 | 64
0.27992848 | 96 | 0 | 0.90625 | 64
0.13900934 | 97 | 0 | 0.953125 | 64
0.27177253 | 98 | 0 | 0.921875 | 64
0.15787081 | 99 | 0 | 0.921875 | 64
0.40764943 | 99 | 1 | 0.890625 | 64
0.12967248 | 99 | 2 | 0.9375 | 64
0.16256529 | 99 | 3 | 0.953125 | 64
0.08198662 | 99 | 4 | 0.96875 | 64
0.12792362 | 99 | 5 | 0.96875 | 64
0.1427121 | 99 | 6 | 0.953125 | 64
0.19214444 | 99 | 7 | 0.953125 | 64
0.26682153 | 99 | 8 | 0.890625 | 64
0.11921781 | 99 | 9 | 0.9375 | 64

 - content of training_set_agg_statistics.csv:

episode | episodes_aggregated | loss | loss_min | loss_max | loss_std | epoch | acc | acc_min | acc_max | acc_std | samples_aggregated
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
99 | 100 | 0.77779132 | 0.09402215 | 2.42056847 | 0.70779961 | 0 | 0.72874999 | 0.015625 | 0.96875 | 0.27987459 | 6400
99 | 1 | 0.40764943 | 0.40764943 | 0.40764943 | 0 | 1 | 0.890625 | 0.890625 | 0.890625 | 0 | 64
99 | 1 | 0.12967248 | 0.12967248 | 0.12967248 | 0 | 2 | 0.9375 | 0.9375 | 0.9375 | 0 | 64
99 | 1 | 0.16256529 | 0.16256529 | 0.16256529 | 0 | 3 | 0.953125 | 0.953125 | 0.953125 | 0 | 64
99 | 1 | 0.08198662 | 0.08198662 | 0.08198662 | 0 | 4 | 0.96875 | 0.96875 | 0.96875 | 0 | 64
99 | 1 | 0.12792362 | 0.12792362 | 0.12792362 | 0 | 5 | 0.96875 | 0.96875 | 0.96875 | 0 | 64
99 | 1 | 0.1427121 | 0.1427121 | 0.1427121 | 0 | 6 | 0.953125 | 0.953125 | 0.953125 | 0 | 64
99 | 1 | 0.19214444 | 0.19214444 | 0.19214444 | 0 | 7 | 0.953125 | 0.953125 | 0.953125 | 0 | 64
99 | 1 | 0.26682153 | 0.26682153 | 0.26682153 | 0 | 8 | 0.890625 | 0.890625 | 0.890625 | 0 | 64
99 | 1 | 0.11921781 | 0.11921781 | 0.11921781 | 0 | 9 | 0.9375 | 0.9375 | 0.9375 | 0 | 64

 - content of validation_statistics.csv:
loss	episode	epoch	acc	batch_size
2.378291607	0	0	0.125	64

 - content of validation_set_agg_statistics.csv:

episode | episodes_aggregated | loss | loss_min | loss_max | loss_std | epoch | acc | acc_min | acc_max | acc_std | samples_aggregated
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
99 | 79 | 0.18435289 | 0.00570344 | 0.51182282 | 0.1067069 | 0 | 0.9398734 | 0.859375 | 1 | 0.03431381 | 5000
99 | 79 | 0.15693755 | 0.03428814 | 0.49671429 | 0.08907631 | 1 | 0.94996047 | 0.84375 | 1 | 0.0299593 | 5000
99 | 79 | 0.13773099 | 0.01599673 | 0.4856357 | 0.07974177 | 2 | 0.95925635 | 0.890625 | 1 | 0.02568884 | 5000
99 | 79 | 0.1349417 | 0.02542676 | 0.38587141 | 0.07309812 | 3 | 0.96004748 | 0.875 | 1 | 0.02536133 | 5000
99 | 79 | 0.13557728 | 0.02810043 | 0.51120263 | 0.07920972 | 4 | 0.95866299 | 0.890625 | 1 | 0.02552018 | 5000
99 | 79 | 0.14610203 | 0.03679127 | 0.45367861 | 0.08219316 | 5 | 0.9535206 | 0.875 | 1 | 0.02991695 | 5000
99 | 79 | 0.14048618 | 0.04181131 | 0.55806714 | 0.08375487 | 6 | 0.95530063 | 0.890625 | 1 | 0.02448055 | 5000
99 | 79 | 0.13374574 | 0.02522391 | 0.46379766 | 0.08368348 | 7 | 0.95787185 | 0.875 | 1 | 0.02812438 | 5000
99 | 79 | 0.13794875 | 0.02302515 | 0.46463043 | 0.07694951 | 8 | 0.9541139 | 0.875 | 1 | 0.0276742 | 5000
99 | 79 | 0.14701881 | 0.00304925 | 0.52258015 | 0.09639523 | 9 | 0.95391613 | 0.890625 | 1 | 0.02705158 | 5000
99 | 79 | 0.14747529 | 0.02490817 | 0.49389082 | 0.08180903 | 9 | 0.95391613 | 0.875 | 1 | 0.02829573 | 5000

### LINKED ISSUE #67: Store termination cause in model checkpoint
- set initial value of "termination_cause" to "[Not Converged]"
 - set termination_cause depending on termination reason
 - pass termination_cause to model.save as additional field

Thanks to this the grid-analyzer will simply inform the user whether model converged or not! 👍

### PR DESCRIPTION
**Pull request by tkornut**
Several fixes of offline/online trainers along the line presented in #66 
In particular:
 - both trainers along with checkpoint export aggregated statistics (ref to #66)
 - Fixes #64 by adding another break in offline trainer
 - Fixes episode increment in offline trainer
 - Fixes #67 stores termination cause in checkpoint (both trainers)

### CODE REVIEW

**Comment by tkornut** [comment]
This pull request **introduces 1 alert** when merging 5b8e77cb79e58b366dd44e488bf57fd0aeb42d75 into 194cedec3431c7b56d1c1b2fed1a6e75e9fec265 - [view on LGTM.com](https://lgtm.com/projects/g/IBM/mi-prometheus/rev/pr-d45f166034057dc8dc83146c586913e40096f5a3)

**new alerts:**

* 1 for Wrong number of arguments in a call

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: miprometheus/models/model.py [hunk 1]
```diff
@@ -317,38 +317,35 @@ def plot(self, data_dict, predictions, sample=0):
...
         :param model_dir: Directory where the model will be saved.
         :type model_dir: str
 
-        :param stats: Statistics value to save with the model.
-        :type stats: ``StatisticsCollector`` or ``StatisticsAggregator``
+        :param training_status: String representing the current status of training.
+        :type training_status: str
+
+        :param stats: Statistics that will be saved to checkpoint along with the model.
+        :type stats: ``StatisticsAggregator``
```
Comment (vmarois) [inline miprometheus/models/model.py]: If you want to specify that the type of `params` is the `StatisticsAggregator` class, can you please use the following ? This will allow inter-linking in the doc :slightly_smiling_face: 

> :type stats: :py:class:`miprometheus.utils.StatisticsAggregator`

File: miprometheus/models/model.py [hunk 2]
```diff
@@ -317,38 +317,35 @@ def plot(self, data_dict, predictions, sample=0):
 
         """
 
-    def save(self, model_dir, stats):
+    def save(self, model_dir, training_status, stats):
```
Comment (vmarois) [inline miprometheus/models/model.py]: So you are adding a new argument to this method, which means that where it is called (with `super()` for instance), it needs to be modified as well...

File: miprometheus/workers/offline_trainer.py [hunk 3]
```diff
@@ -265,18 +267,19 @@ def run_experiment(self):
...
+
                         # Save the model using the latest validation statistics.
-                        self.model.save(self.model_dir, self.validation_stat_col)
+                        self.model.save(self.model_dir, training_status, self.validation_stat_agg)
 
                     # III. The episodes number limit has been reached.
                     if episode+1 >= self.episode_limit:
-                        termination_cause = "Episode Limit reached"
-                        break
+                        training_status = "Episode Limit reached"
```
Comment (vmarois) [inline miprometheus/workers/offline_trainer.py]: So here you are checking `if episode+1 >= self.episode_limit`, but you changed the initial value of `episode`, and where it is incremented in the loop. Does that not cause the training to finish one episode earlier?

File: miprometheus/workers/offline_trainer.py [hunk 4]
```diff
@@ -288,6 +291,11 @@ def run_experiment(self):
                 # Apply curriculum learning - change some of the Problem parameters
                 self.curric_done = self.training_problem.curriculum_learning_update_params(episode)
 
+                # III. The episodes number limit has been reached. (2nd check)
+                if episode+1 >= self.episode_limit:
```
Comment (vmarois) [inline miprometheus/workers/offline_trainer.py]: Why a second check?

File: miprometheus/models/model.py (hunk 2 above)
Comment (vmarois) [reply miprometheus/models/model.py]: LGTM has caught that the call to this function in `MAES.save()` does not have 3 arguments, so it will break

Comment (tkornut) [reply miprometheus/models/model.py]: I know, I saw that as well, fixing...;)

File: miprometheus/workers/offline_trainer.py (hunk 3 above)
Comment (tkornut) [reply miprometheus/workers/offline_trainer.py]: Double checked it in both workers, it is ok.
For example, setting episode_limit = 133 results in collection of statistics up to episode 132, aggregation also happens in episode 132.

File: miprometheus/workers/offline_trainer.py (hunk 4 above)
Comment (tkornut) [reply miprometheus/workers/offline_trainer.py]: Required to break outer (epoch) loop when episode limit reached.

**Comment by tkornut** [comment]
Newest push fixes #74

File: configs/vision/simplecnn_mnist.yaml [hunk 5]
```diff
@@ -19,30 +20,33 @@ training:
...
         episode_limit: 20000
-        epoch_limit: 10
+        epoch_limit: 1
 
 # Problem parameters:
 validation:
     problem:
         name: *name
-        batch_size: *b
+        batch_size: 5000
```
Comment (vmarois) [inline configs/vision/simplecnn_mnist.yaml]: You specify the use of `SubsetRandomSampler` below, so a more 'regular' batch size could be used, no?

File: configs/vision/simplecnn_mnist.yaml [hunk 6]
```diff
@@ -19,30 +20,33 @@ training:
...
+    #    drop_last: True
 
 # Problem parameters:
 testing:
+    #seed_numpy: 4354
+    #seed_torch: 2452
     problem:
         name: *name
-        batch_size: *b
+        batch_size: 10000
```
Comment (vmarois) [inline configs/vision/simplecnn_mnist.yaml]: Same remark here

File: miprometheus/models/encoder_solver/maes_model.py [hunk 7]
```diff
@@ -100,24 +100,33 @@ def __init__(self, params, problem_default_values_={}):
...
-        
-        :param stat_col: Statistics collector that contain current loss and episode number (and other statistics).
+        :type model_dir: str
+
+        :param training_status: String representing the current status of training.
+        :type training_status: str
+
+        :param training_stats: Training statistics that will be saved to checkpoint along with the model.
+        :type training_stats: :py:class:miprometheus.utils.StatisticsAggregator or :py:class:miprometheus.utils.StatisticsAggregator
+
```
Comment (vmarois) [inline miprometheus/models/encoder_solver/maes_model.py]: Small thing: you are missing back ticks and you copy-pasted twice the same class name ;)

File: miprometheus/models/model.py [hunk 8]
```diff
@@ -328,26 +328,47 @@ def save(self, model_dir, training_status, stats):
         :param training_status: String representing the current status of training.
         :type training_status: str
 
-        :param stats: Statistics that will be saved to checkpoint along with the model.
-        :type stats: ``StatisticsAggregator``
+        :param training_stats: Training statistics that will be saved to checkpoint along with the model.
+        :type training_stats: :py:class:miprometheus.utils.StatisticsAggregator or :py:class:miprometheus.utils.StatisticsAggregator
+
```
Comment (vmarois) [inline miprometheus/models/model.py]: Same small remark here

File: miprometheus/workers/offline_trainer.py [hunk 9]
```diff
@@ -271,15 +273,14 @@ def run_experiment(self):
                         self.aggregate_and_export_statistics(self.model, self.validation_problem, 
                                 self.validation_stat_col, self.validation_stat_agg, episode, '[Partial Validation]', False)
 
-                        # Save the model using the latest validation statistics.
-                        self.model.save(self.model_dir, training_status, self.validation_stat_agg)
+                        # Do not save the model! Offline trainer uses full set fot determining whether to save or not.
```
Comment (vmarois) [inline miprometheus/workers/offline_trainer.py]: Yes 👍

File: miprometheus/models/encoder_solver/maes_model.py (hunk 7 above)
Comment (tkornut) [reply miprometheus/models/encoder_solver/maes_model.py]: done :)

File: miprometheus/models/model.py (hunk 1 above)
Comment (tkornut) [reply miprometheus/models/model.py]: 👌
